File: 1001-2000/1575-count-all-possible-routes/1575-count-all-possible-routes.cpp
const int kMod = 1000'000'007;

int ModAdd(int64_t a, int64_t b) {
  return (a + b) % kMod;
}

class Solution {
public:
  int countRoutes(vector<int>& locations, int start, int finish, int fuel) {
    int start_location = locations[start];
    int finish_location = locations[finish];
    std::sort(locations.begin(), locations.end());
    start = std::lower_bound(locations.begin(), locations.end(), start_location) - locations.begin();
    finish = std::lower_bound(locations.begin(), locations.end(), finish_location) - locations.begin();
    
    const int N = locations.size();
    const int M = fuel + 1;
    vector<vector<int>> dpR(N, vector<int>(M));
    vector<vector<int>> dpL(N, vector<int>(M));
    for (int x = 1; x < M; ++x) {
      for (int i = 1; i < N; ++i) {
        int y = x - (locations[i] - locations[i - 1]);
        if (y < 0) {
          continue;
        }
        dpR[i - 1][x] = ModAdd(dpR[i][y] * 2, dpL[i][y]);
        dpL[i][x] = ModAdd(dpL[i - 1][y] * 2, dpR[i - 1][y]);
        if (i == finish) {
          dpR[i - 1][x] = ModAdd(dpR[i - 1][x], 1);
        }
        if (i - 1 == finish) {
          dpL[i][x] = ModAdd(dpL[i][x], 1);
        }
      }
    }
    int answer = dpR[start][fuel] + dpL[start][fuel];
    if (start == finish) {
      ++answer;
    }
    return answer % kMod;
  }
};

File: 1001-2000/1575-count-all-possible-routes/1575-count-all-possible-routes.java
class Solution {
    public int countRoutes(int[] locations, int start, int finish, int fuel) {
        
        int n = locations.length;
        int MOD = 1_000_000_007;
        
        int[][] dp = new int[n][fuel + 1];

        // Start at start city with 0 fuel used
        
        dp[start][0] = 1;

        for (int f = 0; f <= fuel; f++) {
            for (int city = 0; city < n; city++) {

                if (dp[city][f] == 0) continue;

                for (int next = 0; next < n; next++) {

                    if (next == city) continue;

                    int cost = Math.abs(locations[city] - locations[next]);

                    if (f + cost <= fuel) {
                        dp[next][f + cost] =
                                (dp[next][f + cost] + dp[city][f]) % MOD;
                    }
                }
            }
        }

        // Sum all ways reaching finish with <= fuel used
        int ans = 0;
        for (int f = 0; f <= fuel; f++) {
            ans = (ans + dp[finish][f]) % MOD;
        }

        return ans;
    }

    private int dfsMemo(int[] locations, int start, int finish, int fuel,int[][] memo){
        
        if (fuel < 0) return 0;

        if(memo[start][fuel]!=-1)return memo[start][fuel];

        int paths = 0;

        // Count current position if it's finish
        if (start == finish) {
            paths++;
        }

        for(int i=0;i<locations.length;i++){
            if(i!=start){
                int fuelNeed=Math.abs(locations[start]-locations[i]);
                if(fuel-fuelNeed>=0){
                    paths= (paths+dfsMemo(locations,i,finish,fuel-fuelNeed,memo))%1000000007;
                }
            }
        }
        return memo[start][fuel]=paths;
    }
}